src/main/integrations/discord-presence/index.ts

```diff
@@ -78,92 +64,72 @@ function stringLimit(str: string, limit: number, minimum: number) {
+        },
+        {
+          label: "Play on YouTube Music Desktop",
+          url: `ytmd://play/${id}`
         }
-      } else {
-        this.lastDuration = duration;
-        this.lastTimeSeconds = timeSeconds;
-      }
+      ]
+    });
+  }
 
-      this.endTimestamp = state.trackState === VideoState.Playing ? adjustedTimeSeconds + (state.videoDetails.durationSeconds - videoProgress) : undefined;
-      if (
-        this.lastEndTimestamp !== this.endTimestamp ||
-        state.videoDetails.title !== this.lastVideoDetailsTitle ||
-        state.videoDetails.author !== this.lastVideoDetailsAuthor ||
-        state.videoDetails.id !== this.lastVideoDetailsId ||
-        state.trackState !== this.lastTrackState
-      ) {
-        this.lastEndTimestamp = this.endTimestamp;
-        this.lastVideoDetailsTitle = state.videoDetails.title;
-        this.lastVideoDetailsAuthor = state.videoDetails.author;
-        this.lastVideoDetailsId = state.videoDetails.id;
-        this.lastTrackState = state.trackState;
+  private playerStateChanged(state: PlayerState) {
+    if (this.ready && state.videoDetails) {
+      const oldState = this.videoState ?? null;
+
+      this.videoDetails = {
+        title: state.videoDetails.title,
+        author: state.videoDetails.author,
+        album: state.videoDetails.album,
+        id: state.videoDetails.id,
+        thumbnails: state.videoDetails.thumbnails,
+        durationSeconds: state.videoDetails.durationSeconds
+      };
+      this.progress = Math.floor(state.videoProgress);
+      this.videoState = state.trackState;
 
-        const thumbnail = getHighestResThumbnail(state.videoDetails.thumbnails);
-        this.discordClient.setActivity({
-          // Discord accepts Playing, Listening, and Watching. But ignores it and sets it to 0 (Playing)
-          // We'll still send a type in case Discord at some point updates to allow this
-          type: DiscordActivityType.Listening,
-          details: stringLimit(state.videoDetails.title, 128, 2),
-          state: stringLimit(state.videoDetails.author, 128, 2),
-          timestamps: {
-            end: state.trackState === VideoState.Playing ? this.endTimestamp : undefined
-          },
-          assets: {
-            large_image: thumbnail && thumbnail.length <= 256 ? thumbnail : "ytmd-logo",
-            large_text: stringLimit(state.videoDetails.title, 128, 2),
-            small_image: getSmallImageKey(state.trackState),
-            small_text: getSmallImageText(state.trackState)
-          },
-          instance: false,
-          buttons: [
-            {
-              label: "Play on YouTube Music",
-              url: `https://music.youtube.com/watch?v=${state.videoDetails.id}`
-            },
-            {
-              label: "Play on YouTube Music Desktop",
-              url: `ytmd://play/${state.videoDetails.id}`
-            }
-          ]
-        });
+      if (oldState !== this.videoState) {
```

Comment:
Have you tested it yet? This 'simplification' handles everything

Reply:
While it may work today it may not work tomorrow as we don't make nor want to make a guarantee the video state has to change for other data to change. This could cause the activity to display outdated data.

Reply:
The video always buffers when changing songs
If you want to be extra careful you can measure the change in progress on the video to see if it was moved fast or something to be extra careful

Reply:
I don't feel like you understand what he means.
This feature can work today with the current state of YouTube Music.

But we have no idea how YouTube Music's internals might update in the future, where progression between songs / tracks could be instant and never trigger this state to change. This is why we want more things to be considered for track changing.

Reply:
I feel like alot more would break than just this

Reply:
https://github.com/ytmdesktop/ytmdesktop/blob/013ea5b4d6cecf35a49eaef30e92784057728b08/src/main/integrations/notifications/index.ts#L78-L79 - Notifications, Specifically checks that it's only playing, but then checks if the ID has changed.


https://github.com/ytmdesktop/ytmdesktop/blob/013ea5b4d6cecf35a49eaef30e92784057728b08/src/main/index.ts#L604 - Taskbar, Checking for if the track is playing for Progress on the Taskbar, other ones will show in pause state + Quick actions when hovering.


https://github.com/ytmdesktop/ytmdesktop/blob/013ea5b4d6cecf35a49eaef30e92784057728b08/src/main/integrations/last-fm/index.ts#L77-L80 - Last.FM, Checks if playing and from list of posibile IDs (Watching or Counterpart if using Premium) to make sure to not double scrobble.


None of these are checking for a State change so loosely and have other checks to see if the ID of the video has changed or more.
So in my generic example I provided nothing else would break.

Reply:
I added checks for the video details id changing, or progress changing, or state changing to update activity.